numpyro/contrib/distributions/discrete.py

```diff
@@ -1,35 +1,72 @@
+# The implementation largely follows the design in PyTorch's `torch.distributions`
+#
+# Copyright (c) 2016-     Facebook, Inc            (Adam Paszke)
+# Copyright (c) 2014-     Facebook, Inc            (Soumith Chintala)
+# Copyright (c) 2011-2014 Idiap Research Institute (Ronan Collobert)
+# Copyright (c) 2012-2014 Deepmind Technologies    (Koray Kavukcuoglu)
+# Copyright (c) 2011-2012 NEC Laboratories America (Koray Kavukcuoglu)
+# Copyright (c) 2011-2013 NYU                      (Clement Farabet)
+# Copyright (c) 2006-2010 NEC Laboratories America (Ronan Collobert, Leon Bottou, Iain Melvin, Jason Weston)
+# Copyright (c) 2006      Idiap Research Institute (Samy Bengio)
+# Copyright (c) 2001-2004 Idiap Research Institute (Ronan Collobert, Samy Bengio, Johnny Mariethoz)
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+
+
 import jax.numpy as np
 import jax.random as random
 from jax import lax
-from jax.scipy.special import gammaln
+from jax.scipy.special import gammaln, logsumexp
 
 from numpyro.contrib.distributions.distribution import Distribution
 from numpyro.distributions import constraints
 from numpyro.distributions.util import (
     binary_cross_entropy_with_logits,
     binomial,
+    categorical_rvs,
     get_dtypes,
     lazy_property,
+    multinomial_rvs,
     promote_shapes,
     xlog1py,
-    xlogy
-)
+    xlogy,
+    poisson)
 
 
-def _to_probs_bernoulli(logits):
-    return 1 / (1 + np.exp(-logits))
+def clamp_probs(probs):
+    finfo = np.finfo(get_dtypes(probs)[0])
+    return np.clip(probs, a_min=finfo.tiny, a_max=1. - finfo.eps)
```

Comment:
@fehiepsi : That's the reason why I'm using `1. - finfo.eps` for `a_max`. `tiny` shouldn't be used for that - it will result in a `1.` and lead to issues.

Reply:
Sorry, I mean that the issue happens for some operators which requires `1 - p < 1`. But we have already addressed them at StickBreakingTransform (at least). I see no other issues in the log_prob of current distributions.

Reply:
Your observation is right. Note that `tiny` is the smallest representable number with the given dtype. We cannot do `1. - tiny` and expect it to be less than 1. `eps` on the other hand gives the smallest number such that `1 + eps != 1`, and that works also with `1 - eps != 1` (as far as I can tell). That's the reason why `a_min` uses `tiny` but `a_max` uses `eps` so that the result is guaranteed to be in `(0, 1)` (exclusive).

Reply:
Understood!